Edge trait

wks opened the pull request:
We make the representation of edges VM-specific.  The MMTk core only
depends on the abstract Edge trait.  This allows VM bindings to custom
the behaviour of loading/storing of edges, and support features like
compressed OOPs.

Some example `Edge` implementations are in `vmbindings/dummyvm/src/edges.rs`.  There is a test case `vmbindings/dummyvm/src/tests/edges_test.rs` to show what to expect.

This PR involves API change.  Migration guide is in the commit message.

Binding PRs:
-   OpenJDK: https://github.com/mmtk/mmtk-openjdk/pull/160
-   JikesRVM: https://github.com/mmtk/mmtk-jikesrvm/pull/121
-   V8: https://github.com/mmtk/mmtk-v8/pull/66
-   Ruby: https://github.com/mmtk/mmtk-ruby/pull/9

Closes: https://github.com/mmtk/mmtk-core/issues/573

qinsoon commented on src/vm/edge_shape.rs:
```diff
@@ -0,0 +1,115 @@
+/// For backword compatibility.
+/// We let Address implement Edge so existing bindings that use `Address` to represent an edge can
+/// continue to work.
+impl Edge for Address {
```
Maybe it is worth documenting that we suggest using `SimpleEdge` instead of `Address`, and why.

wks replied:
Yes.   I'll add more comments about why we should move away from using Address directly as an edge.

qinsoon commented on vmbindings/dummyvm/src/edges.rs:
```diff
@@ -0,0 +1,157 @@
+
+/// This edge presents the object reference itself to mmtk-core.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct ValueEdge {
```
Will we use a type like this when a binding wants to do node enqueuing? If so, we probably should just call it `Node` or at least in the comment, state this is also known as node.

wks replied:
I thought this `ValueEdge` could help node-enqueuing, but I later found it doesn't help.  I already removed all comments that mentioned "edge enqueuing" in this PR, but I want to keep this example, just to demonstrate that the developers can think out of the box when implementing `Edge`.  I'll add some comments to clarify the purpose of this example.

qinsoon reviewed (approved):
LGTM

qinsoon commented:
I suggest we also get a review from @wenyuzhao on this PR.

qinsoon commented on vmbindings/dummyvm/src/edges.rs:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    fn store(&self, _object: ObjectReference) {
+        // No-op.  Value edges are immutable.
```
I don't think we can call this an edge. We call `store()` to update the edge, but this implementation cannot update. The `store()` impl should be at least `unreachable!()` or `panic!()`, as the impl cannot do what it is asked. However, that will cause MMTk crash, as we will call `store`. I think we should just remove this type -- it is confusing.

wks replied:
OK. I think I'll just remove the `ValueEdge` type because it is confusing.

wenyuzhao commented on src/vm/edge_shape.rs:
```diff
@@ -0,0 +1,115 @@
+/// A simple edge implementation that represents a word-sized slot where an ObjectReference value
+/// is stored as is.  It is the default edge type, and should be suitable for most VMs.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct SimpleEdge {
```
`#[repr(transparent)]` so that the size of the struct is equal to the pointer size.

wks replied:
It doesn't have to.  What matters is its member `slot_addr: *mut Atomic<ObjectReference>,` **points** to a location that has pointer-sized thing.

Edge instances are passed from the VM to mmtk-core in callbacks like `scan_object` and edge-scanning functions.  More often than not, `Edge` instances will *not* have the same size as the pointer.  The `OffsetEdge` type in the [vmbindings/dummyvm/src/edges.rs](https://github.com/mmtk/mmtk-core/pull/606/files/24b7e47fc1647efec38f3d41e9419de619f70d99#diff-e1acc241ac12bd754b4502e45a5b174b3db0360ac50f11fed7f2c79d0e5e4a03) is an example.  An `Edge` instance also holds *information* about an edge, such as offsets, or tag bits, in addition to the address of the field.

wks replied:
Well, I'll just add `#[repr(transparent)]` because some VMs may want to pass field address from C to Rust and want Rust to treat it as an edge.

wenyuzhao commented on src/vm/mod.rs:
```diff
@@ -45,6 +46,9 @@ where
     type VMReferenceGlue: ReferenceGlue<Self>;
 
+    /// The type of edges in this VM.
+    type VMEdge: edge_shape::Edge;
```
Is it possible to do a runtime switch? For example, OpenJDK may need to support compressed pointers as an optional run-time feature, and cannot be activated at build-time.

wenyuzhao replied:
Also probably this can be moved to VMObjectModel?

wks replied:
If the "feature" is a Cargo feature, we can define it like this:
```rust
enum OpenJDKEdge {
    Uncompressed(*mut AtomicUsize),
    #[cfg(features = "compressed_oop")]
    Compressed(*mut AtomicU32),
}
```

But if we want it enabled/disabled by a command-line flag, the easiest way (may affect performance, but need to be confirmed with tests) is simply not use the `OpenJDKEdge::Compressed` case.  As an optimisation, maybe we can parameterise `OpenJDK` to `OpenJDK<EdgeType>` so that depending on the command line options, we can instantiate MMTk with different (parameterised) VMBinding implementation.

qinsoon replied:
Maybe we could check the performance impact of having `enum Edge` (but never using the other case). We do hope that is minimal.

qinsoon replied:
> Also probably this can be moved to VMObjectModel?

I don't think `ObjectModel` is the right place to put it. I think it is fine to put the type in `VMBinding` (as it is now). If we plan to move it, I think `Scanning` fits well. All related changes are in the `Scanning` trait. This is an indication that the type is related with `Scanning`.

wks replied:
> Also probably this can be moved to VMObjectModel?

In theory, we can.  But then instead of `VM::VMEdge`, we will need to write `<<VM as VMBinding>::VMObjectModel as ObjectModel<VM>>::VMEdge` everywhere.

A refactoring may be possible to let VMBinding *require* VMObjectModel, instead of including a VMObjectModel as a member.  I have seen similar problems when I tried to refactor ReferenceProcessor.  I tried to add a function in `VMCollection` that depends on a type in `VMReferenceGlue`, and I have to navigate from VMCollection to VMBinding, and from there to VMReferenceGlue.  Sometimes the compiler is not smart enough to figure out the VM type passed to the function to VMCollection is the same VM as the VM in `VMReferenceGlue<VM>`.

I'll elaborate the refactoring in a different issue.

wenyuzhao replied:
Using `enum OpenJDKEdge` will (1) introduce an extra branch for each edge load and store (2) For a mark-queue `Vec<Edge>`, there will be one word (100%) overhead for each entry. With compressed pointers on, the overhead will be 300%. Probably I'll parameterize `OpenJDK`.

wenyuzhao replied:
> In theory, we can. But then instead of VM::VMEdge, we will need to write <<VM as VMBinding>::VMObjectModel as ObjectModel<VM>>::VMEdge everywhere.

I see. Yes I agree it should stay in VMBinding.

wenyuzhao replied:
I think I don't really need another edge type for compressed pointers. The size of an edge **address** is always one word, no matter whether it's a `*mut AtomicU32` or `*mut AtomicUsize `. The only thing I need to replace is the handler of the load/store operations.

wks replied:
If space overhead is a problem, the OpenJDK binding can implement `Edge` with a tagged pointer.  It saves space, but still needs a test on bits.

Other bindings may not be able to use this trick.  Ruby already uses tagged pointers, and Julia has offsetted edges where the offset has to be recorded.

wks commented:
I updated the PR.
1.  I replaced the `ValueEdge` example with `TaggedEdge`, because we do have some VMs using tagged pointers.
2.  I added comments about why we should prefer `SimpleEdge` over `Address`.
3.  I added `#[repr(transparent)]` to `SimpleEdge`.
